official/coin_build/man/man3/SoDetail.3
.TH "SoDetail" 3 "Sun May 28 2017" "Version 4.0.0a" "Coin" \" -*- nroff -*-
.ad l
.nh
.SH NAME
SoDetail \- The \fBSoDetail\fP class is the superclass for all classes storing detailed information about particular shapes\&.
.PP
Detail information about shapes is used in relation to picking actions in Coin\&. They typically contain the relevant information about what particular part of the shape a pick ray intersected with\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <Inventor/details/SoDetail\&.h>\fP
.PP
Inherited by \fBSoConeDetail\fP, \fBSoCubeDetail\fP, \fBSoCylinderDetail\fP, \fBSoFaceDetail\fP, \fBSoLineDetail\fP, \fBSoNodeKitDetail\fP, \fBSoPointDetail\fP, and \fBSoTextDetail\fP\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "virtual \fB~SoDetail\fP ()"
.br
.ti -1c
.RI "virtual \fBSoDetail\fP * \fBcopy\fP (void) const =0"
.br
.ti -1c
.RI "virtual \fBSoType\fP \fBgetTypeId\fP (void) const =0"
.br
.ti -1c
.RI "SbBool \fBisOfType\fP (const \fBSoType\fP type) const"
.br
.in -1c
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "static void \fBinitClass\fP (void)"
.br
.ti -1c
.RI "static void \fBinitClasses\fP (void)"
.br
.ti -1c
.RI "static \fBSoType\fP \fBgetClassTypeId\fP (void)"
.br
.in -1c
.SS "Protected Member Functions"

.in +1c
.ti -1c
.RI "\fBSoDetail\fP (void)"
.br
.in -1c
.SH "Detailed Description"
.PP 
The \fBSoDetail\fP class is the superclass for all classes storing detailed information about particular shapes\&.
.PP
Detail information about shapes is used in relation to picking actions in Coin\&. They typically contain the relevant information about what particular part of the shape a pick ray intersected with\&. 
.SH "Constructor & Destructor Documentation"
.PP 
.SS "SoDetail::~SoDetail ()\fC [virtual]\fP"
Destructor\&. 
.SS "SoDetail::SoDetail (void)\fC [protected]\fP"
Default constructor\&. 
.SH "Member Function Documentation"
.PP 
.SS "void SoDetail::initClass (void)\fC [static]\fP"
Initialize relevant common data for all instances, like the type system\&. 
.SS "void SoDetail::initClasses (void)\fC [static]\fP"
Call the \fBinitClass()\fP methods of all built-in detail classes\&.
.PP
(The \fBinitClass()\fP method of user extension detail classes -- if any -- must be called explicitly by the application programmer in the application initialization code\&.) 
.SS "\fBSoDetail\fP * SoDetail::copy (void) const\fC [pure virtual]\fP"
Return a deep copy of ourself\&.
.PP
\fBNote that if the Coin library has been built as a DLL under Microsoft Windows and you use this method from application code, you \fImust\fP make sure that both the Coin DLL and the application executable is using the same instance of a C Run-Time (CRT) library\&. Otherwise, you \fIwill\fP get memory heap corruption upon deallocating the returned instances, eventually leading to mysterious crashes\&.\fP 
.PP
Implemented in \fBSoNodeKitDetail\fP, \fBSoLineDetail\fP, \fBSoFaceDetail\fP, \fBSoTextDetail\fP, \fBSoCubeDetail\fP, \fBSoCylinderDetail\fP, \fBSoConeDetail\fP, and \fBSoPointDetail\fP\&.
.SS "\fBSoType\fP SoDetail::getTypeId (void) const\fC [pure virtual]\fP"
Returns the type identification of a detail derived from a class inheriting \fBSoDetail\fP\&. This is used for run-time type checking and 'downward' casting\&.
.PP
Usage example:
.PP
.PP
.nf
void fuhbear(SoDetail * detail)
{
  if (detail->getTypeId() == SoFaceDetail::getClassTypeId()) {
    // safe downward cast, know the type
    SoFaceDetail * facedetail = (SoFaceDetail *)detail;
  }
  return; // ignore if not a SoFaceDetail
}
.fi
.PP
.PP
For application programmers wanting to extend the library with new detail classes: this method needs to be overridden in \fIall\fP subclasses\&. This is typically done as part of setting up the full type system for extension classes, which is usually accomplished by using the pre-defined macros available through Inventor/nodes/SoSubDetail\&.h: SO_DETAIL_SOURCE and SO_DETAIL_INIT_CLASS\&. 
.PP
Implemented in \fBSoNodeKitDetail\fP, \fBSoPointDetail\fP, \fBSoLineDetail\fP, \fBSoFaceDetail\fP, \fBSoTextDetail\fP, \fBSoCubeDetail\fP, \fBSoCylinderDetail\fP, and \fBSoConeDetail\fP\&.
.SS "SbBool SoDetail::isOfType (const \fBSoType\fP type) const"
Returns \fCTRUE\fP if \fItype\fP is derived from (or \fIis\fP) this class\&. 
.SS "\fBSoType\fP SoDetail::getClassTypeId (void)\fC [static]\fP"
Returns the type for this class\&. 

.SH "Author"
.PP 
Generated automatically by Doxygen for Coin from the source code\&.


insanity/coin/src/details/SoDetail.cpp
/**************************************************************************\
 * Copyright (c) Kongsberg Oil & Gas Technologies AS
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 * 
 * Redistributions of source code must retain the above copyright notice,
 * this list of conditions and the following disclaimer.
 * 
 * Redistributions in binary form must reproduce the above copyright
 * notice, this list of conditions and the following disclaimer in the
 * documentation and/or other materials provided with the distribution.
 * 
 * Neither the name of the copyright holder nor the names of its
 * contributors may be used to endorse or promote products derived from
 * this software without specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
\**************************************************************************/

/*!
  \class SoDetail SoDetail.h Inventor/details/SoDetail.h
  \brief The SoDetail class is the superclass for all classes storing
  detailed information about particular shapes.
  \ingroup details

  Detail information about shapes is used in relation to picking
  actions in Coin.  They typically contain the relevant information
  about what particular part of the shape a pick ray intersected with.

*/

// *************************************************************************

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif // HAVE_CONFIG_H

#include <Inventor/SbName.h>
#include <Inventor/details/SoConeDetail.h>
#include <Inventor/details/SoCubeDetail.h>
#include <Inventor/details/SoCylinderDetail.h>
#include <Inventor/details/SoFaceDetail.h>
#include <Inventor/details/SoLineDetail.h>
#include <Inventor/details/SoPointDetail.h>
#include <Inventor/details/SoTextDetail.h>
#ifdef HAVE_NODEKITS
#include <Inventor/details/SoNodeKitDetail.h>
#endif // HAVE_NODEKITS

// *************************************************************************

/*!
  \fn SoDetail * SoDetail::copy(void) const
  Return a deep copy of ourself.

  \DANGEROUS_ALLOC_RETURN
*/

// Note: the following documentation for getTypeId() will also be
// visible for subclasses, so keep it general.  If you write any
// additional documentation for this method, check the other top-level
// classes with getTypeId() documentation to see if it is applicable
// to update those also (it probably is).
/*!
  \fn SoType SoDetail::getTypeId(void) const

  Returns the type identification of a detail derived from a class
  inheriting SoDetail.  This is used for run-time type checking and
  "downward" casting.

  Usage example:

  \code
  void fuhbear(SoDetail * detail)
  {
    if (detail->getTypeId() == SoFaceDetail::getClassTypeId()) {
      // safe downward cast, know the type
      SoFaceDetail * facedetail = (SoFaceDetail *)detail;
      /// [then something] ///
    }
    return; // ignore if not a SoFaceDetail
  }
  \endcode


  For application programmers wanting to extend the library with new
  detail classes: this method needs to be overridden in \e all
  subclasses. This is typically done as part of setting up the full
  type system for extension classes, which is usually accomplished by
  using the pre-defined macros available through
  Inventor/nodes/SoSubDetail.h: SO_DETAIL_SOURCE and
  SO_DETAIL_INIT_CLASS.
*/

// *************************************************************************

SoType SoDetail::classTypeId STATIC_SOTYPE_INIT;

// *************************************************************************

/*!
  Default constructor.
*/
SoDetail::SoDetail(void)
{
}

/*!
  Destructor.
*/
SoDetail::~SoDetail()
{
}

/*!
  Returns \c true if \a type is derived from (or \e is) this class.
*/
bool
SoDetail::isOfType(const SoType type) const
{
  return this->getTypeId().isDerivedFrom(type);
}

/*!
  Returns the type for this class.
*/
SoType
SoDetail::getClassTypeId(void)
{
  return SoDetail::classTypeId;
}

// Note: the following documentation for initClass() will also be
// visible for subclasses, so keep it general.
/*!
  Initialize relevant common data for all instances, like the type
  system.
 */
void
SoDetail::initClass(void)
{
  SoDetail::classTypeId =
    SoType::createType(SoType::badType(), SbName("SoDetail"));
  SoDetail::initClasses();
}

/*!
  Call the initClass() methods of all built-in detail classes.

  (The initClass() method of user extension detail classes -- if any
  -- must be called explicitly by the application programmer in the
  application initialization code.)
 */
void
SoDetail::initClasses(void)
{
  SoConeDetail::initClass();
  SoCubeDetail::initClass();
  SoCylinderDetail::initClass();
  SoFaceDetail::initClass();
  SoLineDetail::initClass();
  SoPointDetail::initClass();
  SoTextDetail::initClass();
#ifdef HAVE_NODEKITS
  SoNodeKitDetail::initClass();
#endif // HAVE_NODEKITS
}
